### PR TITLE
testnet faucet

### LINKED ISSUE #278: testnet faucet

### PR DESCRIPTION
**Pull request by fedekunze**
Closes: #278 

## Description

```bash
emintcli q bank balances $(emintcli keys show mykey -a)    
emintcli tx faucet fund 10photon,1uatom,100aragon --from mykey
emintcli q bank balances $(emintcli keys show mykey -a)
emintcli q faucet funded   
```

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: x/faucet/genesis.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/ethermint/x/faucet/types"
+	abci "github.com/tendermint/tendermint/abci/types"
+)
+
+// InitGenesis initializes genesis state based on exported genesis
+func InitGenesis(ctx sdk.Context, k Keeper, data types.GenesisState) []abci.ValidatorUpdate {
+	if acc := k.GetFaucetAccount(ctx); acc == nil {
+		panic(fmt.Sprintf("%s module account has not been set", ModuleName))
```
Comment (noot) [inline x/faucet/genesis.go:14]: will this affect the daemon running if it panics?

**Review by noot — APPROVED** [review]
looks great! 

I have a few questions, it looks like there is an HTTP server, could you provide an example of querying that server? is it started by default with the `emintcli rest-server` command or does it need to be started separately?

also, when I do `emintcli q faucet funded` in your instructions, I see `[]`, is this expected?

File: x/faucet/genesis.go (hunk 1 above)
Comment (fedekunze) [reply x/faucet/genesis.go:14]: it will panic on when starting the daemon (if the account is not set)

**Comment by fedekunze** [comment]
> I have a few questions, it looks like there is an HTTP server, could you provide an example of querying that server?

So this is what I said during standup one time. Ethermint's REST server was replaced by the EVM Web3 RPC so we can't use the SDK default REST server unless we move that RPC to another process (i.e meaning that we'll have to run 3 processes: daemon, rest server, and the evm module's web3 RPC). This is the same argument why clients (explorers, wallets, etc) cannot use other modules' or general REST HTTP endpoints such as staking or standard txs.

I added the corresponding TODO and I think we should address for the testnet to have at least 1 supported block explorer. Ref #304 @GregTheGreek @ColinSchwarz

**Comment by fedekunze** [comment]
> also, when I do `emintcli q faucet funded` in your instructions, I see [], is this expected?

that's expected when you haven't yet funded any address. Did your tx succeed?

**Comment by noot** [comment]
@fedekunze I saw 
```
I[2020-05-18|15:05:33.572] Added good transaction                       module=mempool tx=CC7E07043B11AF69CBE2B55F5ED3B6E3C5D0F063B6475CDBF6DCCF444A3A59A5 res="&{CheckTx:log:\"[]\" gas_wanted:200000 gas_used:56781 }" height=7 total=1

```

but the balance of my account didn't go down?

**Pull request merged**